packages/client/src/components/ClientProvider/ClientProvider.tsx

```diff
@@ -204,19 +118,70 @@ const ClientProvider = ({
               }
             },
           }),
-          dedupExchange,
           cacheExchange,
-          authExchange({
-            getAuth,
-            addAuthToOperation,
-            didAuthError,
-            willAuthError,
+          authExchange(async (utils) => {
+            const {
+              accessToken,
+              refreshToken,
+              idToken,
+              logout,
+              refreshTokenSet,
+            } = authRef.current;
+
+            return {
+              addAuthToOperation(operation) {
+                if (accessToken) {
+                  return utils.appendHeaders(operation, {
+                    Authorization: `Bearer ${accessToken}`,
+                  });
+                }
+                return operation;
+              },
+              willAuthError: () =>
+                willAuthError({
+                  accessToken,
+                  refreshToken,
+                  idToken,
+                }),
+              didAuthError(error) {
+                return error && error.response
+                  ? error.response.status === 401 ||
+                      error.graphQLErrors.some(
+                        (e) => e.extensions?.category === "authentication",
+                      )
+                  : false;
+              },
+              async refreshAuth() {
```

Comment:
Hmm, I don't think the refresh is working for me.  It looks like the refresh request was properly issued but the subsequent requested didn't use the new toke?
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/767977a3-c687-43f2-9806-da5f7f2d1777)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/300fb9f0-fc9e-4df9-a9e5-793549435082)

Reply:
It looks like you were right about the `useCallback`. I moved it into that and it seems to work. 🤷‍♀️ 

[move auth refresh to useCallback](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6547/commits/dbf68c7eae57611f0c1e9c0c9f0d9a853a244dc7)

Reply:
Nice, the refresh works!  It seems not-quite-right though.  I'm observing it taking two tries before it "sticks".  This is the sort of thing that "willAuthError" is supposed to avoid, I think.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/6993b801-d613-44f2-abfa-8515d7858ba2)

I'm going to try with SIC to see if it does the same thing.

Reply:
Thanks, yeah they replaced it with some `refreshAuth` and said to define it inline. I used what we used to have and merged it with their example code but I am not super sure about what i did 😅

Reply:
Hmm, closer look at the timings.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/1a2c4255-007f-4b44-8a43-8f5c63143d9b)
It's actually attempting the second refresh before it "knows" that the graphql request errored out.  :confused:

Reply:
I wonder if it has something to do with either the `return` or `await`? 🤔 I added the return to satisy the type and no other reason (I don't remember them returning anything from this function)

https://github.com/GCTC-NTGC/gc-digital-talent/blob/dbf68c7eae57611f0c1e9c0c9f0d9a853a244dc7/packages/client/src/components/ClientProvider/ClientProvider.tsx#L86-L93

Reply:
I bet it is the new dependencies of the `internalClient`:
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/915596c3-91d3-4e23-ae81-a8c60ccd1736)
When we get a new refresh token we get a new `internalClient` with a reset state that doesn't know it has requests in flight.
This sounds an awful lot like  #2041 so it's probably a similar problem with a similar solution.  The reason this is dangerous is because if a client keeps getting 401's repeatedly they will get stuck in a refresh loop.  :fearful:

Reply:
Or it's just plain not working.  :laughing: I tried SIC which is more strict, of course.  The refresh just plain fails.  I think the new tokens are not getting loaded into state properly.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/61b934e6-5df7-4ec9-ac48-06ea1d2cacdc)

Reply:
SIC detail:
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/a7c6a31f-2140-4eea-993d-3549b6e5e336)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/15d7992a-e59e-4b1d-bca9-71f1e990e2ae)

Reply:
Hmm, okay. This should give me enough info to work off of. Thanks. I may be pinging you later if I need more assistance 😛

Reply:
Not entirely sure if this works, still triggers two refreshes but no more 401's 😬 

- [move auth refresh to useCallback](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6547/commits/dbf68c7eae57611f0c1e9c0c9f0d9a853a244dc7)
- [update client memo dependencies](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6547/commits/8d6ac47237ce32606e8947f2c124b404add3aeec)

Reply:
OK, I tested again with SIC.  I'm not seeing it try the refresh twice but it still isn't handling the state properly.  It seems to wait until the refresh is done but it tries again with the old token, fails, then tries with the new token.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/74e4bf88-f1fe-4359-b062-c7bba08bd377)